mongofiles/gfsfile.go

```diff
@@ -0,0 +1,152 @@
+package mongofiles
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/mongodb/mongo-go-driver/bson"
+	"github.com/mongodb/mongo-go-driver/bson/primitive"
+	"github.com/mongodb/mongo-go-driver/mongo"
+	"github.com/mongodb/mongo-go-driver/mongo/gridfs"
+	driverOptions "github.com/mongodb/mongo-go-driver/mongo/options"
+	"github.com/mongodb/mongo-go-driver/x/bsonx"
+)
+
+// Struct representing a GridFS files collection document.
+type gfsFile struct {
+	ID         interface{}     `bson:"_id"`
+	Name       string          `bson:"filename"`
+	Length     int64           `bson:"length"`
+	Md5        string          `bson:"md5"`
+	UploadDate time.Time       `bson:"uploadDate"`
+	Metadata   gfsFileMetadata `bson:"metadata"`
+	ChunkSize   int            `bson:"chunkSize"`
+
+	// Storage required for reading and writing GridFS files
+	mf         *MongoFiles
+	downStream *gridfs.DownloadStream
+	upStream   *gridfs.UploadStream
+}
+
+// Struct representing the metadata associated with a GridFS files collection document.
+type gfsFileMetadata struct {
+	ContentType string             `bson:"contentType,omitempty"`
+}
+
+func newGfsFile(ID interface{}, name string, mf *MongoFiles) (*gfsFile, error) {
+	if ID == nil || mf == nil {
+		return nil, fmt.Errorf("invalid gfsFile arguments, one of ID (%v) or MongoFiles (%v) nil", ID, mf)
+	}
+
+	return &gfsFile{Name: name, ID: ID, mf: mf}, nil
+}
+
+func newGfsFileFromCursor(cursor mongo.Cursor, mf *MongoFiles) (*gfsFile, error) {
+	if mf == nil {
+		return nil, fmt.Errorf("invalid gfsFile argument, MongoFiles nil")
+	}
+
+	var out gfsFile
+	if err := cursor.Decode(&out); err != nil {
+		return nil, fmt.Errorf("error decoding GFSFile: %v", err)
+	}
+
+	if out.ID == nil {
+		return nil, fmt.Errorf("invalid gfsFile, ID nil")
+	}
+
+	out.mf = mf
+
+	return &out, nil
+}
+
+// Write data to GridFS Upload Stream. If this file has not been written to before, this function will open a new stream that must be closed.
+// Note: the go driver buffers data until it hits a chunk size before writing to the database, so if the amount of data is written < chunkSize,
+// the actual write to the database will occur in Close.
+func (file *gfsFile) Write(p []byte) (int, error) {
+	if file.upStream == nil {
```

Comment:
The more I reflect on this, the more I think we shouldn't be combining "open" with "write".  Already the "new..." functions are differentiated for reading and writing contexts.  I get that you're trying to anticipate future usage, like with Delete() checking for an upload stream, but I think it winds up confusing comprehension of how the code is used currently.  I think it would be better to have the gfsFile produce upload/download streams as needed and leave responsibility for closing them to the caller rather than wrapping Read/Write.  I understand that on Swift there's a 1-10 scale for -- on this I'm about a 7 out of 10 for being more explicit and less magical.

Reply:
In a more recent commit than the one this comment is on, I added `OpenForWriting` and `OpenForReading` that must occur outside of the `Write` and `Read` functions respectively. Given that, do you think I should still do away with the `Writer`/`Reader` implementation and just provide the streams as needed, or leave as-is?

Reply:
Great!  I think you could trivially have `OpenForWriting/Reading` return the associated stream rather than caching it in the object and then you won't need `Read` and `Write` wrappers.

Reply:
Done. 

This actually leaves `gfsfile.go` a little bare. Is it still worth providing the `Delete`, `OpenForX` functions on the `gfsFile` type, or should I just make them functions of `MongoFiles` instead?